typescript/xstate/src/promise.ts

```diff
@@ -0,0 +1,227 @@
+  input: TInput | undefined;
+  sent: boolean;
+};
+
+const RESTATE_PROMISE_SENT = 'restate.promise.sent';
+const RESTATE_PROMISE_RESOLVE = 'restate.promise.resolve';
+const RESTATE_PROMISE_REJECT = 'restate.promise.reject';
+const XSTATE_STOP = 'xstate.stop';
+
+type PromiseCreator<TOutput, TInput extends NonReducibleUnknown> = ({input, ctx}: {
+  input: TInput;
+  ctx: restate.RpcContext,
+}) => PromiseLike<TOutput>
+
+export type PromiseActorLogic<TOutput, TInput = unknown> = ActorLogic<
+  PromiseSnapshot<TOutput, TInput>,
+  { type: string; [k: string]: unknown },
+  TInput, // input
+  AnyActorSystem
+> & { sentinel: "restate.promise.actor", config: PromiseCreator<TOutput, TInput> };
+
+export type PromiseActorRef<TOutput> = ActorRefFrom<
+  PromiseActorLogic<TOutput, unknown>
+>;
+
+export function fromPromise<TOutput, TInput extends NonReducibleUnknown>(
+  promiseCreator: PromiseCreator<TOutput, TInput>,
+): PromiseActorLogic<TOutput, TInput> {
+  const logic: PromiseActorLogic<TOutput, TInput> = {
+    sentinel: "restate.promise.actor",
+    config: promiseCreator,
+    transition: (state, event) => {
+      if (state.status !== 'active') {
+        return state;
+      }
+
+      switch (event.type) {
+        case RESTATE_PROMISE_SENT: {
+          return {
+            ...state,
+            sent: true,
+          };
+        }
+        case RESTATE_PROMISE_RESOLVE: {
+          const resolvedValue = (event as any).data;
+          return {
+            ...state,
+            status: 'done',
+            output: resolvedValue,
+            input: undefined
+          };
+        }
+        case RESTATE_PROMISE_REJECT:
+          return {
+            ...state,
+            status: 'error',
+            error: (event as any).data,
+            input: undefined
+          };
+        case XSTATE_STOP:
+          return {
+            ...state,
+            status: 'stopped',
+            input: undefined
+          };
+        default:
+          return state;
+      }
+    },
+    start: (state, {self, system}) => {
+      if (state.status !== 'active') {
+        return;
+      }
+
+      if (state.sent) {
+        return;
+      }
+
+      const rs = (system as RestateActorSystem<any>)
+
+      rs.ctx.send(rs.api.promise).invoke({
+        systemName: rs.systemName,
+        self: serialiseActorRef(self),
+        srcs: actorSrc(self),
+        input: state.input
+      });
+
+      // note that we sent off the promise so we don't do it again
+      (system as any)._relay(self, self, {
+        type: RESTATE_PROMISE_SENT,
+      });
+    },
+    getInitialSnapshot: (_, input) => {
+      return {
+        status: 'active',
+        output: undefined,
+        error: undefined,
+        input,
+        sent: false,
+      };
+    },
+    getPersistedSnapshot: (snapshot) => snapshot,
+    restoreSnapshot: (snapshot: any) => snapshot
+  };
+
+  return logic;
+}
+
+function actorSrc(actor?: AnyActorRef): string[] {
```

Comment:
tries to determine the chain of `src` fields which is the best way i could find to derive a 'path' through the state machine definition that helps you find the promise creator